drivers/modem/ublox-sara-r4.c

```diff
@@ -176,1114 +128,516 @@ struct modem_iface_ctx {
-	struct modem_socket *sock = NULL;
-
-	/* make sure only a single digit is picked up for socket_id */
-	value[0] = net_buf_pull_u8(*buf);
-	len--;
-	value[1] = 0;
-
-	socket_id = atoi(value);
-	if (socket_id < MDM_BASE_SOCKET_NUM) {
-		return;
-	}
-
-	/* skip first comma */
-	net_buf_pull_u8(*buf);
-	len--;
-
-	/* Second parameter is length */
-	out_len = net_buf_linearize(value, sizeof(value) - 1, *buf, 0, len);
-	value[out_len] = 0;
-	left_bytes = atoi(value);
+	struct modem_socket *sock;
 
-	sock = socket_from_id(socket_id);
+	sock = modem_socket_from_id(&mdata.socket_config, atoi(argv[0]));
 	if (!sock) {
-		LOG_ERR("Unable to find socket_id:%d", socket_id);
 		return;
 	}
 
-	if (left_bytes > 0) {
-		LOG_DBG("socket_id:%d left_bytes:%d", socket_id, left_bytes);
-
-		snprintk(sendbuf, sizeof(sendbuf), "AT+USOR%s=%d,%d",
-			 (sock->ip_proto == IPPROTO_UDP) ? "F" : "D",
-			 sock->socket_id, left_bytes);
-
-		/* We entered this trigger due to an unsolicited modem response.
-		 * When we send the AT+USOR* command it won't generate an
-		 * "OK" response directly.  The modem will respond with
-		 * "+USOR*: ..." and the data requested and then "OK" or
-		 * "ERROR".  Let's not wait here by passing in a timeout to
-		 * send_at_cmd().  Instead, when the resulting response is
-		 * received, we trigger on_cmd_sockread() to handle it.
-		 */
-		send_at_cmd(sock, sendbuf, K_NO_WAIT);
-	}
-}
-
-static void on_cmd_socknotifycreg(struct net_buf **buf, u16_t len)
-{
-	char value[8];
-	size_t out_len;
-
-	out_len = net_buf_linearize(value, sizeof(value) - 1, *buf, 0, len);
-	value[out_len] = 0;
-	ictx.ev_creg = atoi(value);
-	LOG_DBG("CREG:%d", ictx.ev_creg);
+	modem_socket_put(&mdata.socket_config, sock->sock_fd);
 }
 
-static int net_buf_ncmp(struct net_buf *buf, const u8_t *s2, size_t n)
-{
-	struct net_buf *frag = buf;
-	u16_t offset = 0U;
-
-	while ((n > 0) && (*(frag->data + offset) == *s2) && (*s2 != '\0')) {
-		if (offset == frag->len) {
-			if (!frag->frags) {
-				break;
-			}
-			frag = frag->frags;
-			offset = 0U;
-		} else {
-			offset++;
-		}
-
-		s2++;
-		n--;
-	}
-
-	return (n == 0) ? 0 : (*(frag->data + offset) - *s2);
-}
-
-static inline struct net_buf *read_rx_allocator(s32_t timeout, void *user_data)
-{
-	return net_buf_alloc((struct net_buf_pool *)user_data, timeout);
-}
-
-static void modem_read_rx(struct net_buf **buf)
-{
-	u8_t uart_buffer[MDM_RECV_BUF_SIZE];
-	size_t bytes_read = 0;
-	int ret;
-	u16_t rx_len;
-
-	/* read all of the data from mdm_receiver */
-	while (true) {
-		ret = mdm_receiver_recv(&ictx.mdm_ctx,
-					uart_buffer,
-					sizeof(uart_buffer),
-					&bytes_read);
-		if (ret < 0 || bytes_read == 0) {
-			/* mdm_receiver buffer is empty */
-			break;
-		}
+/* Handler: +UUSOR[D|F]: <socket_id>[0],<length>[1] */
+MODEM_CMD_DEFINE(on_cmd_socknotifydata)
+{
+	int ret, socket_id, new_total;
+	struct modem_socket *sock;
 
-		hexdump(uart_buffer, bytes_read);
+	socket_id = atoi(argv[0]);
```

Comment:
I just wonder if we should use strtol() instead as atoi() cannot figure out errors in the string.
So all the invocations of atoi() could be replaced by
```
	socket_id = strtol(argv[0], &endptr, 10);
	if (*endptr == '\0') {
		/* string ok */
	} else {
		/* not a number */
	}
```

Reply:
Will do.  Thanks for the recommendation.